### ISSUE TITLE
Attention layer to mix the features

### ISSUE THREAD

**Issue by George3d6** [mindsdb/lightwood#179]
Currently the encoded features go into a fully connected first layer of the mixer.

I'm wondering if it *might* be better to have something like an attention layer that can weight each of the mixer features in part.

Few questions I have here:

1. Is attention even the right mechanisms to use here or is it to complicated ?
2. How does one go about implementing an attention layer that works with variable-sized chunks rather than single values ? 

E.g. given a tensor `[1,0.5,0.2,0.9,0.4,1.1,0]` attention layers would process each value in relation to each other value, but what we have as the input into the mixer is more like `[[1,0.5],[0.2,0.9,0.4],[1.1,0]]`, since each encoded feature is a vector of 1 or more values.

Is there a way to generalize the idea behind attention such that it behaves well in this case ?

Since the encoded features are of different sizes and each value in those tensors might be relevant in some different way we can't just apply e.g. softmax to every vector (which seems to be what people are doing for images: https://lilianweng.github.io/lil-log/2018/06/24/attention-attention.html)

3. If attention is not the way to go, what would be another approach that does something similar:

I.e. give a "weight" to every single encoded feature vector based on the value in the other encoded feature vectors + it's own value.

**Comment by maximlopin** [comment]
**Predictions with missing columns using SAN (self-attention network).**

Encoded features, which is a set of vectors of varying sizes should be linearly transformed to vectors of equal size. Now they can be thought of as word embeddings, and the model architecture from the referenced paper (the paper is below) can be applied.

The only difference is we shouldn't use positional encoding, because order of features doesn't matter.

_I'll update this comment later with some other things to consider._

**References**

_Importance of Self-Attention for Sentiment Analysis_
https://www.aclweb.org/anthology/W18-5429.pdf

**Comment by George3d6** [comment]
I'd be curios what Lezcano and or @paxcema would think about this topic.

**Comment by paxcema** [comment]
I think it's an interesting idea, might be particularly useful to avoid representation bias whenever there's certain IRs that dominate the mixer input in terms of dimensions.

Regarding the questions posed:
1. Very straightforward using `torch.nn.MultiHeadAttention()`
2. As far as I understand, not much of an issue because the mixer receives a flattened vector. Pointing to the attention layer which slice of the vector belongs to which column might help, but in theory it should learn as it is just fine.

I've implemented a rough prototype in the `attentive_defnet` branch. I'll benchmark and see whether it helps or not.

**Comment by paxcema** [comment]
Ok, I'm updating this one for any potential Hacktoberfest contributors:

This issue precedes the Lightwood 1.0 refactor, so things have changed quite a bit. However, the original implementation in the above referenced branch (`attentive_defnet`) should still work fine, so I would say this issue can now be framed as:

1. Reintroduce the `LinAttnBlock` class found [here](https://github.com/mindsdb/lightwood/blob/44f21eebe07becd824d5f4081c9ec9aed9ce843d/lightwood/mixers/helpers/attn_default_net.py#L106), adding it in `mixer/helpers/default_net` should be fine.
2. Introduce a keyword argument and all necessary logic so that if said flag (let's call it `use_attention`) is set to True, then the `DefaultNet` object will instantiate a `LinAttnBlock` and use it. For this logic you can also check the link above as an initial reference.
3. Expose the flag so that it can be modified from a Json AI object.


Feel free to ping me or @George3d6 if you're interested in tackling this and you have any questions or comments.

**Comment by paxcema** [comment]
For anyone interested, this would be a feature to implement inside the `Neural` mixer. Once implemented, we can run the benchmarking suite to figure out whether it's a net benefit. Closing for now.
